node/subsystem-util/src/validator_discovery.rs

```diff
@@ -104,6 +105,75 @@ async fn connect_to_authorities<Context: SubsystemContext>(
 	Ok((connected_rx, revoke_tx))
 }
 
+/// A struct that assists performing multiple concurrent connection requests.
+///
+/// This allows concurrent connections to validator sets at different `relay_parents`
+/// and multiplexes their results into a single `Stream`.
+#[derive(Default)]
+pub struct ConnectionRequests {
+	// added connection requests relay_parent -> StreamUnordered token
+	id_map: HashMap<Hash, usize>,
+
+	// Connection requests themselves.
+	requests: StreamUnordered<ConnectionRequest>,
+}
+
+impl ConnectionRequests {
+	/// Insert a new connection request.
+	///
+	/// If a `ConnectionRequest` under a given `relay_parent` already exists it will
+	/// be revoked and substituted with a new one.
+	pub fn put(&mut self, relay_parent: Hash, request: ConnectionRequest) {
+		self.remove(&relay_parent);
+		let token = self.requests.push(request);
+
+		self.id_map.insert(relay_parent, token);
+	}
+
+	/// Remove a connection request by a given `relay_parent`.
+	pub fn remove(&mut self, relay_parent: &Hash) {
+		if let Some(token) = self.id_map.remove(relay_parent) {
+			Pin::new(&mut self.requests).remove(token);
+		}
+	}
+}
+
+impl stream::Stream for ConnectionRequests {
+	/// (relay_parent, validator_id, peer_id).
+	type Item = (Hash, ValidatorId, PeerId);
+
+	fn poll_next(mut self: Pin<&mut Self>, cx: &mut task::Context) -> Poll<Option<Self::Item>> {
+		loop {
+			match Pin::new(&mut self.requests).poll_next(cx) {
+				Poll::Ready(Some((yeild, token))) => {
+					match yeild {
+						StreamYield::Item(item) => {
+							let relay_parent = self.id_map.iter()
+								.find_map(|(relay_parent, &val)|
+									if val == token {
+										Some(relay_parent)
+									} else {
+										None
+									}
+								)
+								.expect("All requests' streams should be accounted for; qed");
+
+							return Poll::Ready(Some((*relay_parent, item.0, item.1)));
+						},
+						StreamYield::Finished(finished_stream) => {
+							finished_stream.remove(Pin::new(&mut self.requests));
+
+							self.id_map.retain(|_, t| *t != token);
+						}
+					}
+				},
+				Poll::Ready(None) => return Poll::Ready(None),
```

Comment:
I think the way collator is written in #1955, it assumes an infinite stream.

Reply:
No it will just ignore the stream ending. As long as the stream does not panic on polling it again, it should be fine.

Reply:
Well you cannot just simply throw away this `Poll::Ready(None)` and return a `Poll::Pending` here. I think that a check for `is_empty()` is ok though

Reply:
If a stream has ended, and then we push some requests into it, will it be polled again?

Reply:
In my code? Yes.

The structure in my pr wasn't meant to be responsible to wake up the task. If we generalize this, we need to make sure what we do.

Reply:
>  If a stream has ended, and then we push some requests into it, will it be polled again?

what? which?

if the `StreamsUnordered` is empty and it is polled then it returns `Poll::Ready(None)`. to avoid that we should just check for `is_empty()`?

Reply:
Well, yeah. But checking `is_empty` you made the `Poll::Ready(None)` branch unreachable. My point was we should never return `Poll::Ready(None)` here, which we now don't, but it's not clear from reading the code.